include/real/const_precision_iterator.hpp

```diff
@@ -0,0 +1,391 @@
+#ifndef BOOST_CONST_PRECISION_ITERATOR_HPP
+#define BOOST_CONST_PRECISION_ITERATOR_HPP
+
+#include "real/boundary_helper.hpp"
+#include "real/interval.hpp"
+#include "real/real_explicit.hpp"
+#include "real/real_algorithm.hpp"
+#include "real/real_operation.hpp"
+#include "real/real_helpers.hpp"
+#include <memory>
+#include <variant>
+#include <assert.h>
+#include <iterator>
+
+namespace boost {
+    namespace real{
+    /**
+     * @file The const_precision_iterator provides the functionality to iterate through precision intervals
+     * of all three kinds of reals
+     * 
+     * @note variant and visit/visitors are used extensively in this implementation
+     * @TODO: consider doing traversals to lower the demands of recursion
+     * @sa documention on std::variant, std::visit
+     */
+        template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+        template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+        // fwd decl
+        class real;
+        class operation_helpers;
+
+        class const_precision_iterator {
+            public:
+            /**
+             * @brief Determines the maximum precision to use
+             */
+            static std::optional<unsigned int> maximum_precision;
+
+            /// @TODO look into STL-style iterators
+            // typedef std::forward_iterator_tag iterator_category;
+            // typedef void difference_type (?);
+            // typedef ??? value_type
+            // typedef const value_type& reference (?) not necessary because const
+            // typedef const value_type* pointer
+
+            private:
+                typedef std::variant<std::monostate, real_explicit, real_algorithm, real_operation> real_number;
+
+                // needed as friend so that we can update operands' precision itrs
+                friend class operation_helpers;
+                
+                friend class real;
+                friend std::ostream& operator<<(std::ostream& os, const real r);
+                /**
+                 * @brief this holds a ptr to explicit number, algorithmic number, or real_operation
+                 */ 
+                // raw pointer here is ok, precision iterator is always attached to the variant it points to
+                // (refer to real_data.hpp). Any time the variant is destroyed, so is this pointer.
+                real_number * _real_ptr;
```

Comment:
I could switch to a weak pointer, but I'm not sure why.

Reply:
Raw pointer may dangle without any check.

Reply:
There should never be dangling because if the *_real_ptr destroyed, then that means the real_data the real_number is in was destroyed, and hence, the precision iterator (member of real_data) and this pointer would be destroyed as well.